src/modm/math/proportional_unsigned.hpp

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2022, Thomas Sommer
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+#pragma once
+
+#include <algorithm>
+#include <concepts>
+
+#include <modm/math/utils/integer_traits.hpp>
+#include <modm/architecture/interface/assert.hpp>
+
+namespace modm {
+
+/**
+ * @brief 		Unsigned integer with arbitrary number of digits and proportional conversion.
+ * 				Artihmetic over- and underflows are being wrapped, just like builtin integral types.
+ *
+ * @tparam D 	Number of Digits
+ *
+ * @author		Thomas Sommer
+ * @ingroup		modm_math
+ */
+template<int D>
+requires (D > 0)
+class ProportionalUnsigned {
+public:
+	static constexpr int digits = D;
+
+	using T = uint_t<D>::least;
+	static constexpr T bitmask = ::modm::bitmask<D>();
+
+	static constexpr T min = 0;
+	static constexpr T max = bitmask;
+
+	/// constructors
+	constexpr ProportionalUnsigned() = default;
+
+	constexpr ProportionalUnsigned(T value) {
+		#ifdef MODM_DEBUG_BUILD
+		if(not modm_assert_continue_ignore_debug(value <= max, "propun.ctor", "value exceeded max", max)) {
+		 	value_ = std::min(max, value);
+			return;
+		}
+		#endif
+
+		value_ = value;
+	}
+
+	template <int E>
+	requires (D <= E)
+	constexpr ProportionalUnsigned(const ProportionalUnsigned<E>& other)
+		: value_(other.value_ >> (E - D))
+	{}
+
+	template <int E>
+	requires (D > E)
+	constexpr ProportionalUnsigned(const ProportionalUnsigned<E>& other)
+		: value_(other.value_ * max / other.max)
+	{}
+
+	/// assignment operators
+	void operator=(T value) {
+		#ifdef MODM_DEBUG_BUILD
+		if(not modm_assert_continue_ignore_debug(value <= max, "propun.assign", "value exceeded max", max)) {
+		 	value_ = std::min(max, value);
+			return;
+		}
+		#endif
+
+		value_ = value;
+	}
+
+	template <int E>
+	requires (D <= E)
+	constexpr void operator=(const ProportionalUnsigned<E>& other) {
+		value_ = other.value_ >> (E - D);
+	}
+
+	template <int E>
+	requires (D > E)
+	constexpr void operator=(const ProportionalUnsigned<E>& other) {
+		value_ = other.value_ * max / other.max;
+	}
+
+	constexpr T value() const { return value_; }
+	constexpr T& value() { return value_; }
+
+	bool operator<=>(const ProportionalUnsigned& other) const = default;
+
+	/// operator +, -, *, /
+	template<std::integral U>
+	ProportionalUnsigned
+	operator+(U value)
+	{
+		return {T(value_ + value) % (ProportionalUnsigned::max + 1)};
+	}
+
+	template<std::integral U>
+	ProportionalUnsigned
+	operator-(U value)
+	{
+		return {T(value_ - value) % (ProportionalUnsigned::max + 1)};
```

Comment:
The maximum values are all `2^n - 1`. Therefore, the slow modulus operation can be expressed as a bit-and (`& ProportionalUnsigned::max`). The compiler should perform the optimization anyway but I would write it in terms of `&` to make really sure it works even with `-O0`.

Reply:
Good point. Thank you!